src/NuGet.Core/NuGet.Commands/SourcesCommands/SourcesRunner.cs

```diff
@@ -0,0 +1,299 @@
+                                Strings.SourcesCommandAllNameIsReserved));
+            }
+            if (string.IsNullOrEmpty(args.Source))
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandSourceRequired));
+            }
+
+            // Make sure that the Source given is a valid one.
+            if (!PathValidator.IsValidSource(args.Source))
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                 Strings.SourcesCommandInvalidSource));
+            }
+
+            ValidateCredentials(args);
+
+            // Check to see if we already have a registered source with the same name or source
+            var existingSourceWithName = args.SourceProvider.GetPackageSourceByName(args.Name);
+            if (existingSourceWithName != null)
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandUniqueName));
+            }
+            var existingSourceWithSource = args.SourceProvider.GetPackageSourceBySource(args.Source);
+            if (existingSourceWithSource != null)
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandUniqueSource));
+            }
+
+            var newPackageSource = new Configuration.PackageSource(args.Source, args.Name);
+
+            if (!string.IsNullOrEmpty(args.Username))
+            {
+                var credentials = Configuration.PackageSourceCredential.FromUserInput(
+                    args.Name,
+                    args.Username,
+                    args.Password,
+                    args.StorePasswordInClearText,
+                    args.ValidAuthenticationTypes);
+                newPackageSource.Credentials = credentials;
+            }
+
+            args.SourceProvider.AddPackageSource(newPackageSource);
+            args.LogMinimal(string.Format(CultureInfo.CurrentCulture,
+                    Strings.SourcesCommandSourceAddedSuccessfully, args.Name));
+        }
+
+        private static void UpdatePackageSource(SourcesArgs args)
+        {
+            if (string.IsNullOrEmpty(args.Name))
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandNameRequired));
+            }
+
+            var existingSource = args.SourceProvider.GetPackageSourceByName(args.Name);
+            if (existingSource == null)
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandNoMatchingSourcesFound, args.Name));
+            }
+
+            if (!string.IsNullOrEmpty(args.Source) && !existingSource.Source.Equals(args.Source, StringComparison.OrdinalIgnoreCase))
```

Comment:
What about an else here to log (at least for high verbosity) that the source already existed and was unchanged?

Reply:
boilerplate reply: the code in SourcesRunner.cs was moved from SourcesCommand.cs - I have tried to make the minimal changes to it. That said, I'll review each of these comments. thx.

Reply:
nice idea, but not as part of this change. perhaps when i go add some source warnings in a future release.